Sources/DeepLearning/Layer.swift

```diff
@@ -927,6 +983,62 @@ public struct AvgPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
     }
 }
 
+/// An average pooling layer for spatial or spatio-temporal data.
+@_fixed_layout
+public struct AvgPool3D<Scalar: TensorFlowFloatingPoint>: Layer {
+    /// The size of the sliding reduction window for pooling.
+    @noDerivative let poolSize: (Int, Int, Int, Int, Int)
+    /// The strides of the sliding window for each dimension of a 5-D input.
+    /// Strides in non-spatial dimensions must be `1`.
+    @noDerivative let strides: (Int, Int, Int, Int, Int)
+    /// The padding algorithm for pooling.
+    @noDerivative let padding: Padding
+
+    /// Creates an average pooling layer.
+    public init(
+        poolSize: (Int, Int, Int, Int, Int),
+        strides: (Int, Int, Int, Int, Int),
+        padding: Padding
+    ) {
+        self.poolSize = poolSize
+        self.strides = strides
+        self.padding = padding
+    }
+
+    /// Creates an average pooling layer.
+    ///
+    /// - Parameters:
+    ///   - poolSize: Vertical and horizontal factors by which to downscale.
+    ///   - strides: The strides.
+    ///   - padding: The padding.
+    public init(poolSize: (Int, Int, Int), strides: (Int, Int, Int), padding: Padding = .valid) {
+        self.poolSize = (1, poolSize.0, poolSize.1, poolSize.2, 1)
+        self.strides = (1, strides.0, strides.1, strides.2, 1)
+        self.padding = padding
+    }
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return input.averagePooled(kernelSize: poolSize, strides: strides, padding: padding)
+    }
+}
+
+public extension AvgPool3D {
+  /// Creates a AvgPool3D Layer with the specified poolsize and strides. The poolsize
+  /// and strides are square windows.
+  init(poolSize: Int, stride: Int, padding: Padding = .valid)
```

Comment:
Please be mindful of the following:
- Use 4-space indentation.
- Wrap `{` onto the previous line.
- Local variable bindings like `let poolsize = poolSize` are not meaningful, use the argument `poolSize` directly.

I hope those points make sense! Being mindful of them helps us merge your PR more quickly.

Please use the following:
```swift
public extension AvgPool3D {
    /// Creates an average pooling layer with the specified pooling window size and stride. The
    /// pooling window is square.
    init(poolSize: Int, stride: Int, padding: Padding = .valid) {
        self.init(poolSize: (poolSize, poolSize, poolSize),
                  strides: (stride, stride, stride),
                  padding: padding)
    }
}
```

Reply:
Yes, it does. I will pay heed from the next time, Sorry!

Reply:
Also @dan-zheng can you think of any other variations of inits and should i extend an init for 2d max and average pooling as well?

Reply:
Note that this API is not comprehensive: both pooling window sizes and strides can either be specified as "multiple integers" or "one integer, implicitly duplicated".

In other words, all of the following may be useful:
- `init(poolSize: (Int, Int, Int), strides: (Int, Int, Int), padding: Padding)`
- `init(poolSize: Int, strides: Int, padding: Padding)`
- `init(poolSize: (Int, Int, Int), strides: Int, padding: Padding)`
- `init(poolSize: Int, strides: (Int, Int, Int) padding: Padding)`

Something to think about and visit after this PR.
Here's Keras (note "integer or tuple of two integers"):

![image](https://user-images.githubusercontent.com/5590046/56820590-3251fa00-6801-11e9-8c80-01178aec8683.png)

Reply:
While writing the code for this, i thought about exactly this case, but had no idea how to address it exactly. In python, It is easier to handle. I'm not very well versed with init overloading in swift

Reply:
> Yes, it does. I will pay heed from the next time, Sorry!

Thanks a ton! That would be a huge help. 🙂

> Also @dan-zheng can you think of any other variations of inits and should i extend an init for 2d max and average pooling as well?

I just commented above about this.
To match Keras, we could have the following combinations:
- `poolSize`: `(Int, Int, Int)` or `Int`
- `strides`: `(Int, Int, Int)` or `Int?` (if `nil`, then use `poolSize` as strides)